rlog/2024-12-23-merkle-tree.mdx

```diff
@@ -0,0 +1,214 @@
+
+We can take advantage of the sparse nature of SMT to provide shorten proofs.
+Specifically, it is unlikely for entries to cluster together.
+As such, it is efficient to maintain a list of values:
+
+|Null values |
+|---|
+|$d_0 := \mathsf{Hash(null)},$|
+|$d_1 := \mathsf{Hash(d_0 \|\| d_0)},$|
+|$d_2 := \mathsf{Hash(d_1 \|\| d_1)},$|
+|$\vdots$|
+|$d_{n-1} := \mathsf{Hash(d_{n-2} \|\| d_{n-2})}.$|
+
+Each of the $d_i$'s represents the root hash of a Merkle tree with $2^i$ leaves containing $\mathsf{null}$.
+These values can be used to shorten the time to construct a SMT, and the length of proofs.
+
+### Proof of nonmembership
+
+In the first Vac-101 [[5](https://vac.dev/rlog/vac101-membership-with-bloom-filters-and-cuckoo-filters)], we examined Bloom and Cuckoo filters that could be used for proof of membership and nonmembership.
+However, the proof of membership may result in false-positives due to collisions.
+This would affect nonmembership proofs as well.
+Sparse Merkle trees can be adapted to provide greater assurance that a given piece of data is not a member of the digest.
+
+Why is sorting essential?
+The sorting mechanism of data can be arbitrarily chosen.
+However, it is essential that there are no gaps in the ordering.
+It must be known what the maximal number of elements that could ever exist in the digest is.
+A simple method for this is to use a hash function to provide fingerprints to the data.
+Each hash using either SHA-256 or Keccak has 256-bits.
+Our entire digest could consist of a maximum of $2^{256}$ entries.
+This assumes that our digest does not contain collisions.
+
+The fingerprint of a piece of data $\ell$ indicates which leaf of the SMT it is contained in.
+This means that a nonmembership of $\ell$ in the SMT becomes a matter of proving that $\mathsf{null}$ is contained in $\ell$'s location. 
+
+It is crucial for the SMT to be sorted. 
+Otherwise, an malicious party can append the entry $\ell$ to a random location.
+This allows for the malicious party to contradictory proofs that prove both membership and nonmemberhip.
+We note that the requirement that a SMT is sorted can be too strong of an assumption in centralized cases.
+However, sorted is a necessary property of SMT for decentralized systems.
+
+### Verkle Trees
+A proof of membership grows in length as the Merkle tree grows.
+The most obvious approach to remedy this scalability issue is to use Merkle trees where each node has more than 2 children.
+However, this does not fix the issue.
+A proof of membership in a $k$-nary Merkle tree (each node has $k$ children) has a proof size $\log_k(n)(k-1)$.
+The multiple $k-1$ is the number of silbings that a node has on each layer.
+Hence, the scaling is worst than logarithmic.
+For additional information on $k$-nary Merkle trees see John Kuszmaul's paper on Verkle trees [[6](https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf)]
```

Comment:
This feels a bit strange to introduce the paper on Verkle trees before introducing them.